In [1]:
import pandas as pd
titanic = pd.read_csv("data/titanic.csv")
titanic.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

# ¿Cómo manipular datos textuales?

### Escriba todos los caracteres del nombre en minúsculas.

In [2]:
titanic["Name"].str.lower()

0                                braund, mr. owen harris
1      cumings, mrs. john bradley (florence briggs th...
2                                 heikkinen, miss. laina
3           futrelle, mrs. jacques heath (lily may peel)
4                               allen, mr. william henry
                             ...                        
886                                montvila, rev. juozas
887                         graham, miss. margaret edith
888             johnston, miss. catherine helen "carrie"
889                                behr, mr. karl howell
890                                  dooley, mr. patrick
Name: Name, Length: 891, dtype: object

*Observaciones*

> Para poner cada una de las cadenas en la columna Nombre en minúsculas, seleccione la columna Nombre (consulte el tutorial sobre la selección de datos), agregue el descriptor de acceso str y aplique el método inferior. Como tal, cada una de las cadenas se convierte por elementos.

De manera similar a los objetos de fecha y hora en el tutorial de series temporales que tienen un descriptor de acceso dt, una serie de métodos de cadena especializados están disponibles cuando se utiliza el descriptor de acceso str. Estos métodos tienen en general nombres coincidentes con los métodos de cadena integrados equivalentes para elementos individuales, pero se aplican por elementos (¿recuerda los cálculos por elementos?) en cada uno de los valores de las columnas.

### Cree una nueva columna Apellido que contenga el apellido de los pasajeros extrayendo la parte anterior a la coma.

In [3]:
titanic["Name"].str.split(",")

0                             [Braund,  Mr. Owen Harris]
1      [Cumings,  Mrs. John Bradley (Florence Briggs ...
2                              [Heikkinen,  Miss. Laina]
3        [Futrelle,  Mrs. Jacques Heath (Lily May Peel)]
4                            [Allen,  Mr. William Henry]
                             ...                        
886                             [Montvila,  Rev. Juozas]
887                      [Graham,  Miss. Margaret Edith]
888          [Johnston,  Miss. Catherine Helen "Carrie"]
889                             [Behr,  Mr. Karl Howell]
890                               [Dooley,  Mr. Patrick]
Name: Name, Length: 891, dtype: object

*Observaciones*

Usando el método Series.str.split(), cada uno de los valores se devuelve como una lista de 2 elementos. El primer elemento es la parte anterior a la coma y el segundo elemento es la parte posterior a la coma.

In [4]:
titanic["surname"] = titanic["Name"].str.split(",").str.get(0)
titanic["surname"]

0         Braund
1        Cumings
2      Heikkinen
3       Futrelle
4          Allen
         ...    
886     Montvila
887       Graham
888     Johnston
889         Behr
890       Dooley
Name: surname, Length: 891, dtype: object

*Observaciones*

Como solo estamos interesados ​​en la primera parte que representa el apellido (elemento 0), podemos usar nuevamente el descriptor de acceso str y aplicar Series.str.get() para extraer la parte relevante. De hecho, estas funciones de cadena se pueden concatenar para combinar varias funciones a la vez.

### Extrae los datos de los pasajeros de las condesas a bordo del Titanic.

In [9]:
titanic["Name"].str.contains("Countess")

0      False
1      False
2      False
3      False
4      False
       ...  
886    False
887    False
888    False
889    False
890    False
Name: Name, Length: 891, dtype: bool

In [8]:
titanic[titanic["Name"].str.contains("Countess")]

PassengerId  Survived  Pclass  \
759          760         1       1   

                                                  Name     Sex   Age  SibSp  \
759  Rothes, the Countess. of (Lucy Noel Martha Dye...  female  33.0      0   

     Parch  Ticket  Fare Cabin Embarked surname  
759      0  110152  86.5   B77        S  Rothes

*Observaciones*

El método de cadena Series.str.contains() verifica cada uno de los valores en la columna Nombre si la cadena contiene la palabra Condesa y devuelve para cada uno de los valores Verdadero (Condesa es parte del nombre) o Falso (Condesa no es parte del nombre). Esta salida se puede usar para subseleccionar los datos usando la indexación condicional (booleana) introducida en el tutorial de creación de subconjuntos de datos. Como solo había una condesa en el Titanic, obtenemos como resultado una fila.

> Nota: Se admiten extracciones más potentes en cadenas, ya que los métodos Series.str.contains() y Series.str.extract() aceptan expresiones regulares, pero están fuera del alcance de este tutorial.

### ¿Qué pasajero del Titanic tiene el nombre más largo?

In [10]:
titanic["Name"].str.len()

0      23
1      51
2      22
3      44
4      24
       ..
886    21
887    28
888    40
889    21
890    19
Name: Name, Length: 891, dtype: int64

*Observaciones*

Para obtener el nombre más largo, primero debemos obtener las longitudes de cada uno de los nombres en la columna Nombre. Mediante el uso de métodos de cadena de pandas, la función Series.str.len() se aplica a cada uno de los nombres de forma individual (por elementos).

In [12]:
titanic["Name"].str.len().idxmax()

307

*Observaciones*

A continuación, necesitamos obtener la ubicación correspondiente, preferiblemente la etiqueta de índice, en la tabla para la que la longitud del nombre es mayor. El método idxmax() hace exactamente eso. No es un método de cadena y se aplica a números enteros, por lo que no se usa str.

In [13]:
titanic.loc[titanic["Name"].str.len().idxmax(), "Name"]

'Penasco y Castellana, Mrs. Victor de Satode (Maria Josefa Perez de Soto y Vallejo)'

*Observaciones*

Basándonos en el nombre de índice de la fila (307) y la columna (Nombre), podemos hacer una selección usando el operador loc, introducido en el tutorial sobre creación de subconjuntos.

### En la columna "Sexo", reemplace los valores de "masculino" por "M" y los valores de "femenino" por "F".

In [14]:
titanic["Sex_short"] = titanic["Sex"].replace(
    {
        "male": "M",
        "female": "F"
    }
)

titanic["Sex_short"]

0      M
1      F
2      F
3      F
4      M
      ..
886    M
887    F
888    F
889    M
890    M
Name: Sex_short, Length: 891, dtype: object

*Observaciones*

Mientras que replace() no es un método de cadena, proporciona una forma conveniente de usar asignaciones o vocabularios para traducir ciertos valores. Requiere un diccionario para definir el mapeo {desde: hasta}.

> Advertencia:
También hay un método replace() disponible para reemplazar un conjunto específico de caracteres. Sin embargo, al tener un mapeo de múltiples valores, esto se convertiría en:

```
titanic["Sex_short"] = titanic["Sex"].str.replace("female", "F")
titanic["Sex_short"] = titanic["Sex_short"].str.replace("male", "M")
```
>Esto se volvería engorroso y conduciría fácilmente a errores. Solo piense (o pruebe usted mismo) lo que sucedería si esas dos declaraciones se aplicaran en el orden opuesto...

## RECORDAR
* Los métodos de cadena están disponibles usando el descriptor de acceso str.

* Los métodos de cadena funcionan por elementos y se pueden usar para la indexación condicional.

* El método replace es un método conveniente para convertir valores de acuerdo con un diccionario dado.